cloudinit/url_helper.py

```diff
@@ -442,18 +516,82 @@ def timeup(max_wait, start_time):
             return False
         return (max_wait <= 0) or (time.time() - start_time > max_wait)
 
-    loop_n = 0
-    response = None
-    while True:
-        if sleep_time_cb is not None:
-            sleep_time = sleep_time_cb(response, loop_n)
+    def read_url_handle_response(response, url):
+        if not response.contents:
+            reason = "empty response [%s]" % (response.code)
+            url_exc = UrlError(
+                ValueError(reason),
+                code=response.code,
+                headers=response.headers,
+                url=url,
+            )
+        elif not response.ok():
+            reason = "bad status code [%s]" % (response.code)
+            url_exc = UrlError(
+                ValueError(reason),
+                code=response.code,
+                headers=response.headers,
+                url=url,
+            )
         else:
-            sleep_time = int(loop_n / 5) + 1
+            reason = ""
+            url_exc = None
+        return (url_exc, reason)
+
+    def read_url_handle_exceptions(url_reader_cb, urls):
+        reason = ""
+        url = None
+        try:
+            url, response = url_reader_cb(urls)
+            url_exc, reason = read_url_handle_response(response, url)
+            if not url_exc:
+                return (url, response.contents)
+        except UrlError as e:
+            reason = "request error [%s]" % e
+            url_exc = e
+        except Exception as e:
+            reason = "unexpected error [%s]" % e
+            url_exc = e
+        time_taken = int(time.time() - start_time)
+        max_wait_str = "%ss" % max_wait if max_wait else "unlimited"
+        status_msg = "Calling '%s' failed [%s/%s]: %s" % (
+            url,
+            time_taken,
+            max_wait_str,
+            reason,
+        )
+        status_cb(status_msg)
+        if exception_cb:
+            # This can be used to alter the headers that will be sent
+            # in the future, for example this is what the MAAS datasource
+            # does.
+            exception_cb(msg=status_msg, exception=url_exc)
+
+    def read_url_cb(url):
+        return readurl(
+            url,
+            headers={} if headers_cb is None else headers_cb(url),
+            headers_redact=headers_redact,
+            timeout=timeout,
+            check_status=False,
+            request_method=request_method,
+        )
+
+    def read_url_serial():
+        """iterate over list of urls, request each one and handle responses
+        and thrown exceptions individually per url
+        """
+
+        def url_reader_serial(url):
+            return (url, read_url_cb(url))
+
+        nonlocal timeout
```

Comment:
I think the nonlocal here is a smell. If I follow it through, there's nothing dangerous about it, but it'd be easier to reason about if timeout was passed to the function and then passed again to `read_url_cb`.

Reply:
Sounds good to me